src/plugins/compressors/sz_threadsafe_plugin.cc

```diff
@@ -0,0 +1,424 @@
+    pressio_options sz_threadsafe_metrics;
+#if HAVE_WRITESTATS
+    set(sz_threadsafe_metrics, "sz_threadsafe:use_mean", sz_threadsafe_stat.use_mean);
+    set(sz_threadsafe_metrics, "sz_threadsafe:block_size", (unsigned int)sz_threadsafe_stat.blockSize);
+    set(sz_threadsafe_metrics, "sz_threadsafe:lorenzo_blocks", (unsigned int)sz_threadsafe_stat.lorenzoBlocks);
+    set(sz_threadsafe_metrics, "sz_threadsafe:regression_blocks", (unsigned int)sz_threadsafe_stat.regressionBlocks);
+    set(sz_threadsafe_metrics, "sz_threadsafe:total_blocks", (unsigned int)sz_threadsafe_stat.totalBlocks);
+    set(sz_threadsafe_metrics, "sz_threadsafe:huffman_tree_size", (unsigned int)sz_threadsafe_stat.huffmanTreeSize);
+    set(sz_threadsafe_metrics, "sz_threadsafe:huffman_coding_size", (unsigned int)sz_threadsafe_stat.huffmanCodingSize);
+    set(sz_threadsafe_metrics, "sz_threadsafe:huffman_node_count", (unsigned int)sz_threadsafe_stat.huffmanNodeCount);
+    set(sz_threadsafe_metrics, "sz_threadsafe:unpredict_count", (unsigned int)sz_threadsafe_stat.unpredictCount);
+
+    set(sz_threadsafe_metrics, "sz_threadsafe:lorenzo_percent", sz_threadsafe_stat.lorenzoPercent);
+    set(sz_threadsafe_metrics, "sz_threadsafe:regression_percent", sz_threadsafe_stat.lorenzoPercent);
+    set(sz_threadsafe_metrics, "sz_threadsafe:huffman_compression_ratio", sz_threadsafe_stat.huffmanCompressionRatio);
+#endif
+    return sz_threadsafe_metrics;
+  }
+
+  std::shared_ptr<libpressio_compressor_plugin> clone() override {
+	return compat::make_unique<sz_threadsafe_plugin>(*this);
+	//return compressor_plugins().build("sz");
+  }
+
+
+  private:
+  static int libpressio_type_to_sz_threadsafe_type(pressio_dtype type) {
+    switch(type)
+    {
+      case pressio_float_dtype:  return SZ_FLOAT;
+      case pressio_double_dtype: return SZ_DOUBLE;
+      case pressio_uint8_dtype: return SZ_UINT8;
+      case pressio_int8_dtype: return SZ_INT8;
+      case pressio_uint16_dtype: return SZ_UINT16;
+      case pressio_int16_dtype: return SZ_INT16;
+      case pressio_uint32_dtype: return SZ_UINT32;
+      case pressio_int32_dtype: return SZ_INT32;
+      case pressio_uint64_dtype: return SZ_UINT64;
+      case pressio_int64_dtype: return SZ_INT64;
+      case pressio_byte_dtype: return SZ_INT8;
+    }
+    return -1;
+  }
+  std::string sz_version;
+  std::string app = "SZ";
+  void* user_params = &threadsafe_params;
+#if PRESSIO_SZ_VERSION_GREATEREQ(2,1,11,1)
+  pressio_data exafel_peaks_segs;
+  pressio_data exafel_peaks_rows;
+  pressio_data exafel_peaks_cols;
+  pressio_data exafel_calibPanel;
+  exafelSZ_params exafel_params{};
+#endif
+};
+
+std::unique_ptr<libpressio_compressor_plugin> make_c_sz_threadsafe() {
+  return compat::make_unique<sz_threadsafe_plugin>();
+}
+
+static pressio_register compressor_sz_threadsafe_plugin(compressor_plugins(), "sz_threadsafe", [](){ static auto sz_threadsafe = std::make_shared<sz_threadsafe_plugin>(); return sz_threadsafe; });
```

Comment:
This is the wrong behavior.  This returns the same plugin on each creation, which is not what we want.  Instead we want a new instance each time. Look how this is done for ZFP.